src/builder-utils.c

```diff
@@ -534,998 +534,75 @@ builder_gobject_from_data (GType       gtype,
-            }
-        }
-    }
-
-  return TRUE;
-}
-
-static const char *
-strptr (Elf_Scn **scns, GElf_Shdr *shdr, int sec, off_t offset)
-{
-  Elf_Scn *scn;
-  Elf_Data *data;
-
-  scn = scns[sec];
-  if (offset >= 0 && (GElf_Addr) offset < shdr[sec].sh_size)
-    {
-      data = NULL;
-      while ((data = elf_rawdata (scn, data)) != NULL)
-        {
-          if (data->d_buf &&
-              offset >= data->d_off &&
-              offset < data->d_off + data->d_size)
-            return (const char *) data->d_buf + (offset - data->d_off);
-        }
-    }
-
-  return NULL;
-}
-
 char **
 builder_get_debuginfo_file_references (const char *filename, GError **error)
 {
-  Elf *elf = NULL;
-  GElf_Ehdr ehdr;
-  int i, j;
-  glnx_fd_close int fd = -1;
-  DebuginfoData data = { 0 };
-  g_autofree GElf_Shdr *shdr = NULL;
-  g_autofree Elf_Scn **scns = NULL;
-  debug_section_t *debug_sections;
-
-  g_autoptr(GHashTable) files = NULL;
-  char **res;
-
-  fd = open (filename, O_RDONLY);
-  if (fd == -1)
-    {
-      glnx_set_error_from_errno (error);
-      return NULL;
-    }
+  g_autofree char *tmp_path = NULL;
+  glnx_autofd int tmp_fd = -1;
+  g_autoptr(GSubprocess) subp = NULL;
+  g_autoptr(GInputStream) input_stream = NULL;
+  g_autoptr(GDataInputStream) data_stream = NULL;
+  g_autoptr(GPtrArray) files = NULL;
+  const char *debugedit = NULL;
 
-  elf = elf_begin (fd, ELF_C_RDWR_MMAP, NULL);
-  if (elf == NULL)
+  tmp_path = g_build_filename (g_get_tmp_dir (), "flatpak-debugedit-list.XXXXXX", NULL);
+  tmp_fd = g_mkstemp (tmp_path);
+  if (tmp_fd == -1)
     {
-      flatpak_fail (error, "cannot open ELF file: %s", elf_errmsg (-1));
+      glnx_set_prefix_error_from_errno(error, "Creating temp file %s failed", tmp_path);
       return NULL;
     }
 
-  if (elf_kind (elf) != ELF_K_ELF)
-    {
-      flatpak_fail (error, "\"%s\" is not an ELF file", filename);
-      return NULL;
-    }
+  debugedit = g_getenv ("FLATPAK_BUILDER_DEBUGEDIT");
+  if (debugedit == NULL)
+    debugedit = DEBUGEDIT;
 
-  if (gelf_getehdr (elf, &ehdr) == NULL)
-    {
-      flatpak_fail (error, "cannot get the ELF header: %s", elf_errmsg (-1));
-      return NULL;
-    }
+  const char * argv[] = { debugedit, "-l", tmp_path, filename, NULL };
 
-  if (ehdr.e_type != ET_DYN && ehdr.e_type != ET_EXEC && ehdr.e_type != ET_REL)
+  subp = g_subprocess_newv (argv, G_SUBPROCESS_FLAGS_NONE, error);
+  if (subp == NULL || !g_subprocess_wait_check (subp, NULL, error))
     {
-      flatpak_fail (error, "\"%s\" is not a shared library", filename);
+      glnx_prefix_error (error, "Running debugedit failed");
       return NULL;
     }
 
-  elf_flagelf (elf, ELF_C_SET, ELF_F_LAYOUT);
-
-  shdr = g_new0 (GElf_Shdr, ehdr.e_shnum);
-  scns =  g_new0 (Elf_Scn *, ehdr.e_shnum);
+  input_stream = g_unix_input_stream_new (tmp_fd, FALSE);
+  data_stream = g_data_input_stream_new (input_stream);
+  files = g_ptr_array_new_with_free_func (g_free);
 
-  for (i = 0; i < ehdr.e_shnum; ++i)
```

Comment:
At this point you should unlink the tempfile so that it will be gone once tmp_fd is closed. Otherwise we will leak temp files on disk.